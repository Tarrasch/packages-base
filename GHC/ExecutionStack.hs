-----------------------------------------------------------------------------
-- |
-- Module      :  GHC.ExecutionStack
-- Copyright   :  (c) The University of Glasgow 2013
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  cvs-ghc@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC Extensions)
--
-- This is a module for the efficient but inaccurate Stack Traces. If you
-- can take a factor 2 of performance penalty. You should consider using
-- "GHC.Stack" as the stack traces will be more accurate and detailed.
--
-- @
-- myFunction :: IO ()
-- myFunction = do
--      stack <- reifyStack
--      dumpStack stack
-- @
--
-- An 'ExecutionStack' is a data wrapper around 'ByteArray#'. The Array is
-- a reified stack. Each element can be thought as the Instruction
-- Pointers. For languages like C you can see from the instruction pointer
-- where you are in the original Haskell function. The ExecutionStack as
-- described by the STG will only contain pointers to entry code of Info
-- Tables.
--
-----------------------------------------------------------------------------

{-# LANGUAGE UnboxedTuples, MagicHash, RecordWildCards #-}
module GHC.ExecutionStack (
  -- * Simple interface
    reifyStack
  , dumpStack
  -- * Complicated interface
  -- ** ExecutionStack
  , ExecutionStack(..)
  , stackSize
  , stackIndex
  -- ** Retrieve candidates
  , getSimpleTraces
  , getTraces
  , getLocationInfos
  , getNumLocationInfos
  , getAllLocationInfos
  -- ** Other
  , LocationInfo(..)
  , showLocationInfo
  , DwarfUnit(..)
  , dwarfInit
  , dwarfFree
  , dwarfLookupIp
  ) where

import GHC.IO (IO(..)
             , unsafePerformIO)
import GHC.Prim
import GHC.Exts
import Unsafe.Coerce
import System.Mem
import Data.Char
import Data.Int
import Data.Word
import GHC.Word
import Control.Monad (forM_)
import Data.Maybe (listToMaybe)
import Foreign.C.String
import Text.Printf (printf)

-- We include MachDeps for the constants containing the sizes of values
#include "MachDeps.h"

data ExecutionStack = ExecutionStack { unExecutionStack ::  ByteArray# }

instance Show ExecutionStack where
    show = showExecutionStack

showExecutionStack :: ExecutionStack -> String
showExecutionStack stack =
    "Stack trace:\n" ++
    concatMap display ([1..] `zip` traces)
  where
    traces = getSimpleTraces stack
    display (ix, trace) = printf "%4u: " (ix :: Int) ++ displayTrace trace ++ "\n"
    displayTrace (Just li)  = showLocationInfo li
    displayTrace Nothing    = "... unknown ..."

-- | Reify the stack. This is the only way to get an ExecutionStack value.
reifyStack :: IO (ExecutionStack)
reifyStack = IO (\s -> let (# new_s, byteArray# #) = reifyStack# s
                           ba = ExecutionStack byteArray#
                       in (# new_s, ba #) )

-- | Pretty print the stack. Will print it to stdout. Note that this is
-- more efficent than doing 'print' as no intermediete Haskell values will
-- get created
dumpStack :: ExecutionStack -> IO ()
dumpStack (ExecutionStack ba) = IO (\s -> let new_s = dumpStack# ba s
                                          in (# new_s, () #))

-- | You never need to call this. Running it twice has no effect.
dwarfInit :: IO ()
dwarfInit = do
    -- putStrLn "Initializing dwarf"
    IO $ \s -> let s' = dwarfInit# s
               in (# s', () #)

-- | Free the memory allocated when doing 'dwarfInit'. This module doesn't
-- automatically free the memory. Instead it will hang around for the whole
-- program execution once it's initialized.
--
-- You never need to call this. Running it twice has no effect.
--
-- Be careful! The CStrings in 'Locationinfo' will become invalidated!
dwarfFree :: IO ()
dwarfFree = do
    -- putStrLn "Deinitializing dwarf"
    IO $ \s -> let s' = dwarfFree# s
               in (# s', () #)

-- | To use this type in the rest of this module is a TODO
data DwarfUnit = DwarfUnit Addr#

-- | Given the address of a function. Collect data about it
dwarfLookupIp :: Addr#
              -> MutableByteArray# RealWorld {-^ Data will be written here.
                                             Can be of any size, if it's
                                             too small nothing will be
                                             written to it. -}
              -> IO (Int, DwarfUnit) -- ^ (maxNumInfos, dwarfUnit)
dwarfLookupIp addr# mba# = IO $ \s ->
    let (# s', numWritten#, dwarfUnit# #) = dwarfLookupIp# addr# mba# s
    in (# s', (I# numWritten#, DwarfUnit dwarfUnit#) #)

-- | The number of functions on your stack
stackSize :: ExecutionStack -> Int
stackSize stack =
    I# (sizeofByteArray# (unExecutionStack stack)) `div` SIZEOF_HSPTR

stackIndex :: ExecutionStack -> Int -> Addr#
stackIndex (ExecutionStack ba#) (I# i#) = indexAddrArray# ba# i#

-- Arash starts to get angry at #-kinded values..., wish I could use
-- primitive package ...
--
-- Apparently I'm not the first guy, look in
-- libraries/binary/src/Data/Binary/Class.hs
--
-- wrapper for ByteArray# (Arash stole from Class.hs)
data ByteArray = BA   { unBA :: ByteArray# }
data MBA       = MBA  { unMBA :: MutableByteArray# RealWorld }

newByteArray :: Int# -> IO MBA
newByteArray sz = IO $ \s ->
  case newByteArray# sz s of { (# s', arr #) ->
  (# s', MBA arr #) }

freezeByteArray :: MutableByteArray# RealWorld -> IO ByteArray
freezeByteArray arr = IO $ \s ->
  case unsafeFreezeByteArray# arr s of { (# s', arr' #) ->
  (# s', BA arr' #) }

-- | Look up the eventual LocationInfos for the functions in the Stack
-- Trace.
getSimpleTraces :: ExecutionStack
                -> [Maybe LocationInfo]
getSimpleTraces stack = map listToMaybe $ getTraces stack 1

-- | Lookup up LocationInfos of a stack with the accuracy provided
getTraces :: ExecutionStack
          -> Int -- ^ Maximum number of traces. Should be >= 1
          -> [[LocationInfo]] -- ^ For each function, its alternative
getTraces stack maxNumInfos = map (\i -> getLocationInfos stack i maxNumInfos)
                                  [0..(stackSize stack)-1]

-- | Given which stack entry, collect a few candidates for the function.
-- The fewer candidates the faster. We've used a heuristic. We guess that
-- the better candidates are always the first elements in the return
-- value.
getLocationInfos :: ExecutionStack
                 -> Int -- ^ Index in the ExecutionStack. Should be in @[0..stackSize)@
                 -> Int -- ^ Maximum number of candidates Should be positive
                 -> [LocationInfo] -- ^ Candidates
getLocationInfos stack i maxNumInfos = infos
    where
        arraySize = maxNumInfos * sizeofLocationInfo
        infos = unsafePerformIO $ do
          mba <- newByteArray (unI# arraySize)
          (numInfos, _du) <- dwarfLookupIp (stackIndex stack i) (unMBA mba)
          ba <- freezeByteArray (unMBA mba)
          let getInfo ix = indexLocationInfoArray (unBA ba) ix
          return $ map getInfo [0..(numInfos - 1)]

-- | Calculate the number of candidates of source level functions for an
-- Instruction Pointer
--
-- O(n) time complexity, where n is the result.
getNumLocationInfos :: ExecutionStack
                    -> Int -- ^ Index in the Executionstack. Should be in @[0..stackSize)@
                    -> Int -- ^ Total number of candidates
getNumLocationInfos = error "Ok, this one requires some additional C function to be implemented"

getAllLocationInfos :: ExecutionStack
                    -> Int -- ^ Index in the Executionstack. Should be in @[0..stackSize)@
                    -> [LocationInfo] -- ^ All candidates
getAllLocationInfos = error "Something involving getNumLocationInfos"

-- | This is a candidate for a Instruction Pointe.  This struct
-- matches the C struct @DebugInfo_@, from dwarf.h
data LocationInfo = LocationInfo {
           startLine    :: !Word16,
           startCol     :: !Word16,
           endLine      :: !Word16,
           endCol       :: !Word16,
           fileName     :: !CString,
           functionName :: !CString
           }
           deriving(Eq)

-- | Will only work if there have not been any 'dwarfFree' since the value
-- was created. For this reason 'LocationInfo' have no Show instance.
showLocationInfo :: LocationInfo -> String
showLocationInfo LocationInfo{..} =
    peek functionName ++
    " (at " ++ peek fileName ++
    ":" ++ show startLine ++ ":" ++ show startCol ++
    "-" ++ show endLine ++ ":" ++ show endCol ++
    ")"
    -- example: "bindIO (at libraries/base/GHC/Base.lhs:609:61-609:77)"
  where
    peek :: CString -> String
    peek = unsafePerformIO . peekCString
      -- It's safe to read here since the strings are allocated on the
      -- actual C heap, which is not handled by the rts.

indexLocationInfoArray :: ByteArray# -> Int -> LocationInfo
indexLocationInfoArray arr# ix = LocationInfo {..}
  where
    offset = ix * sizeofLocationInfo
    depthIn :: Int -> Int -> Int#
    depthIn sizeInBytes bytesIn = unI# ((offset + bytesIn) `div` sizeInBytes)

    startLine = W16# (indexWord16Array# arr# (depthIn 2 0))
    startCol  = W16# (indexWord16Array# arr# (depthIn 2 2))
    endLine   = W16# (indexWord16Array# arr# (depthIn 2 4))
    endCol    = W16# (indexWord16Array# arr# (depthIn 2 6))

    fileName = Ptr (indexAddrArray# arr# (depthIn SIZEOF_HSPTR 8))
    functionName = Ptr (
      indexAddrArray# arr# (depthIn SIZEOF_HSPTR (8 + SIZEOF_HSPTR)))

sizeofLocationInfo :: Int
sizeofLocationInfo =  4*SIZEOF_WORD16 -- {s,e}{line,column}
                    + SIZEOF_HSPTR -- filename
                    + SIZEOF_HSPTR -- functionname
                    + SIZEOF_HSWORD -- depth

unI# :: Int -> Int#
unI# (I# n#) = n#
