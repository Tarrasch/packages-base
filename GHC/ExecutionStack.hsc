-----------------------------------------------------------------------------
-- |
-- Module      :  GHC.ExecutionStack
-- Copyright   :  (c) The University of Glasgow 2013-2014
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  cvs-ghc@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC Extensions)
--
-- This is a module for the efficient but inaccurate stack traces. If you
-- can take a factor 2 of performance penalty, you should consider using
-- "GHC.Stack" as the stack traces will be more accurate and detailed.
--
-- @
-- myFunction :: IO ()
-- myFunction = do
--      stack <- currentExecutionStack
--      printExecutionStack stack
-- @
--
-- An 'ExecutionStack' is a data wrapper around 'ByteArray#'. The
-- Array is a reified stack. Each element is a code address. For most
-- frames this will be the return address for the stack frame, but
-- for an update frame the address is the entry code of the thunk (if
-- available)
--
-- /Since: 4.7.0.0/
-----------------------------------------------------------------------------

{-# LANGUAGE UnboxedTuples, MagicHash, RecordWildCards #-}
module GHC.ExecutionStack (
  -- * Simple interface
    currentExecutionStack
  , printExecutionStack
  -- * Intermediate interface
  , ExecutionStack ()
  , getStackFrames
  , StackFrame(..)
  , LocationInfo(..)
  , printExecutionStackRts
  -- * Complicated interface
  -- ** Managed loading/unloading
  , dwarfIncRef
  , dwarfDecRef
  , dwarfTryUnload
  -- ** Forceful loading/unloading
  , dwarfForceLoad
  , dwarfForceUnload
  -- ** Destructing `ExecutionStack`
  , stackSize
  , stackIndex
  , stackIndices
  -- ** Looking up instructions
  , getStackFramesNoSync
  , getStackFrame
  , getStackFrameCustom
  , getStackFrameNoSync
  , getStackFrameCustomNoSync
  ) where

import GHC.IO (IO(..))
import GHC.Exts
import GHC.Word (Word16)
import Foreign.C.String (peekCString, CString)
import Foreign.C.Types (CInt (..))
import Foreign.Ptr (Ptr, nullPtr)
import Foreign.Storable (Storable(..))
import Foreign.Marshal (alloca, allocaArray)
import Text.Printf (printf)
import Control.Exception.Base (bracket_)

#include "Rts.h"

data ExecutionStack = ExecutionStack
    { unExecutionStack ::  ByteArray##
    }

-- | The number of functions on your stack
stackSize :: ExecutionStack -> Int
stackSize stack =
    I## (sizeofByteArray## (unExecutionStack stack)) `div` (#const SIZEOF_HSPTR)

stackIndex :: ExecutionStack -> Int -> Ptr Instruction
stackIndex (ExecutionStack ba##) (I## i##) = Ptr (indexAddrArray## ba## i##)

stackIndices :: ExecutionStack -> [Ptr Instruction]
stackIndices stack = map (stackIndex stack) [0..(stackSize stack)-1]

data StackFrame = StackFrame
    { unitName      :: !String -- ^ From symbol table
    , procedureName :: !String -- ^ Also from symbol table
    , locationInfos :: ![LocationInfo] -- ^ Empty without @-g@ flag to @ghc@
    }
  -- Looking at Dwarf.h, this is one DwarfUnit and many DebugInfos.

instance Show StackFrame where
    show = showStackFrame

-- | Location in source code.
--
-- TODO(arash, peter): We should aim to make all the Word16s into proper Ints.
data LocationInfo = LocationInfo
    { startLine    :: !Word16
    , startCol     :: !Word16
    , endLine      :: !Word16
    , endCol       :: !Word16
    , fileName     :: !String
    , functionName :: !String
    }
    deriving(Eq)
  -- This struct matches the C struct @DebugInfo_@, from Dwarf.h

instance Show LocationInfo where
    show = showLocationInfo

-- TODO: Better name than prepareStackFrame?

-- | Like 'show', without @unlines@
prepareStackFrame :: StackFrame -> [String]
prepareStackFrame su | null (locationInfos su) =
        [procedureName su ++ " (using " ++ unitName su ++ ")"]
    --  example: mySrcFun (using /path/lib.so)
prepareStackFrame su | otherwise = map showLocationInfo $ locationInfos su

showStackFrame :: StackFrame -> String
showStackFrame = unlines . prepareStackFrame

-- | Pretty print the execution stack to stdout
printExecutionStack :: ExecutionStack -> IO ()
printExecutionStack stack = do
    frames <- getStackFrames stack
    putStrLn $ displayFramesWithHeader frames

displayFramesWithHeader :: [StackFrame] -> String
displayFramesWithHeader frames =
    "Stack trace (Haskell):\n" ++
    concatMap (uncurry displayFrame) ([0..] `zip` frames)

-- | How one StackFrame is displayed in one stack trace
displayFrame :: Int -> StackFrame -> String
displayFrame ix frame = unlines $ zipWith ($) formatters strings
      where formatters = (printf "%4u: %s" (ix :: Int)) : repeat ("      " ++)
            strings    = prepareStackFrame frame

showLocationInfo :: LocationInfo -> String
showLocationInfo LocationInfo{..} =
    functionName ++
    " (at " ++ fileName ++
    ":" ++ show startLine ++ ":" ++ show startCol ++
    "-" ++ show endLine ++ ":" ++ show endCol ++
    ")"
    -- example: "bindIO (at libraries/base/GHC/Base.lhs:609:61-609:77)"

instance Storable LocationInfo where
    sizeOf _ = (#size struct DebugInfo_)
    alignment _ = alignment (undefined :: CInt)

    peek ptr = do
      startLine    <- #{peek struct DebugInfo_, sline } ptr
      startCol     <- #{peek struct DebugInfo_, scol  } ptr
      endLine      <- #{peek struct DebugInfo_, eline } ptr
      endCol       <- #{peek struct DebugInfo_, ecol  } ptr
      fileName     <- #{peek struct DebugInfo_, file  } ptr >>= peekCString
      functionName <- #{peek struct DebugInfo_, name  } ptr >>= peekCString
      return LocationInfo {..}

    poke ptr (LocationInfo{..}) =
        error "Sorry, we're not really Storable, just use it for peek :("

-- We use these three empty data declarations for type-safety
data DwarfUnit
data DwarfProc
data Instruction

peekDwarfUnitName :: Ptr DwarfUnit -> IO CString
peekDwarfUnitName ptr = #{peek struct DwarfUnit_, name } ptr

peekDwarfProcName :: Ptr DwarfProc -> IO CString
peekDwarfProcName ptr = #{peek struct DwarfProc_, name } ptr

-- | Reify the stack. This is the only way to get an ExecutionStack value.
currentExecutionStack :: IO (ExecutionStack)
currentExecutionStack =
    IO (\s -> let (## new_s, byteArray## ##) = reifyStack## s
                  ba = ExecutionStack byteArray##
              in (## new_s, ba ##) )

-- | Pretty print the stack. Will print it to stdout. Note that this
-- function is faster since it is implemented directly in the RTS.
printExecutionStackRts :: ExecutionStack -> IO ()
printExecutionStackRts (ExecutionStack ba) =
    IO (\s -> let new_s = dumpStack## ba s
              in (## new_s, () ##))

-- | Tell the dwarf module that you want to use dwarf. Synchronized.
foreign import ccall "Dwarf.h dwarf_inc_ref" dwarfIncRef :: IO ()

-- | Tell the dwarf module that you are done using dwarf. Synchronized.
foreign import ccall "Dwarf.h dwarf_dec_ref" dwarfDecRef :: IO ()

-- | Ask the dwarf module if it can unload and free up memory. It will not be
-- able to if the module is in use or if data is not loaded. Synchronized.
--
-- Returns True if dwarf data was unloaded.
foreign import ccall "Dwarf.h dwarf_try_unload" dwarfTryUnload :: IO Bool

foreign import ccall "Dwarf.h dwarf_force_load" dwarfForceLoad :: IO ()

foreign import ccall "Dwarf.h dwarf_force_unload" dwarfForceUnload :: IO ()

-- | For the given instruction pointer, how many LocationInfos does it have?
--
-- Dwarf module must be loaded to use this!
foreign import ccall "Dwarf.h dwarf_addr_num_infos"
    dwarfAddrNumInfos :: Ptr Instruction -> IO CInt

-- | Ask dwarf module for
--
-- Dwarf module must be loaded to use this!
foreign import ccall "Dwarf.h dwarf_lookup_ip"
    dwarfLookupIp ::
       Ptr Instruction -- ^ Code address you want information about
    -> Ptr (Ptr DwarfProc) -- ^ Out: DwarfProc Pointer Pointer
    -> Ptr (Ptr DwarfUnit) -- ^ Out: DwarfUnit Pointer Pointer
    -> Ptr LocationInfo -- ^ Out: LocationInfos to write (array)
    -> CInt -- ^ Max amount of LocationInfo it should write (capacity of array)
    -> IO CInt -- ^ How many LocationInfos was actually written

inDwarf :: IO a -> IO a
inDwarf = bracket_ dwarfIncRef dwarfDecRef

getStackFrameCustom ::
       Ptr Instruction -- ^ Instruction Pointer
    -> Int -- ^ Max amount to write
    -> IO StackFrame -- ^ Result
getStackFrameCustom ip maxNumInfos =
    inDwarf $ getStackFrameCustomNoSync ip maxNumInfos

getStackFrameCustomNoSync ::
       Ptr Instruction -- ^ Instruction Pointer
    -> Int -- ^ Max amount to write
    -> IO StackFrame -- ^ Result
getStackFrameCustomNoSync ip maxNumInfos = do
    alloca $ \ppDwarfProc -> do
      poke ppDwarfProc nullPtr
      alloca $ \ppDwarfUnit ->
        allocaArray maxNumInfos $ \infos -> do
          numWritten <- dwarfLookupIp ip
                                      ppDwarfProc
                                      ppDwarfUnit
                                      infos
                                      cMaxNumInfos
          pDwarfProc <- peek ppDwarfProc
          pDwarfUnit <- peek ppDwarfUnit
          unitName <- stringPeekWith peekDwarfUnitName pDwarfUnit
          procedureName <- stringPeekWith peekDwarfProcName pDwarfProc
          locationInfos <- mapM (peekElemOff infos)
                                [0..(fromIntegral numWritten)-1 :: Int]
          return StackFrame{..}
  where
    cMaxNumInfos = fromIntegral maxNumInfos

-- Note: if you grepped your way to the string "<Data not found>,
-- you probably forgot to compile that module with the `-g` flag to ghc.
stringPeekWith :: (Ptr a -> IO CString) -> Ptr a -> IO String
stringPeekWith peeker ptr | ptr == nullPtr = return "<Data not found>"
stringPeekWith peeker ptr | otherwise      = peeker ptr >>= peekCString

getStackFrame :: Ptr Instruction -> IO StackFrame
getStackFrame ip = inDwarf $ getStackFrameNoSync ip

getStackFrameNoSync :: Ptr Instruction -> IO StackFrame
getStackFrameNoSync ip =
    dwarfAddrNumInfos ip >>= (getStackFrameCustomNoSync ip . fromIntegral)

getStackFrames :: ExecutionStack -> IO [StackFrame]
getStackFrames stack = inDwarf $ getStackFramesNoSync stack

getStackFramesNoSync :: ExecutionStack -> IO [StackFrame]
getStackFramesNoSync = mapM getStackFrameNoSync . stackIndices
