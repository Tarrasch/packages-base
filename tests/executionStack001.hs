{-# LANGUAGE UnboxedTuples, MagicHash #-}
module Main where

import GHC.ExecutionStack
import GHC.IO (IO (..))
import Data.List (isPrefixOf, isSuffixOf)

main = peculiarFunctionName 5 >>= print

checkSensibility :: [StackFrame] -> String
checkSensibility units = str1 ++ str2
    where
        infos = concatMap locationInfos units
        foundPeculiar = any isPeculiar infos
        str1 | foundPeculiar = "Found the peculiar function.\n"
             | otherwise     = "Didn't find the function. :(\n"
        str2 = ""


isPeculiar :: LocationInfo -> Bool
isPeculiar locationInfo = True
        && slin startLine
        && scol startCol
        && elin endLine
        && ecol endCol
        && ("peculiarFunctionName" `isPrefixOf` functionName locationInfo)
        && ("executionStack001.hs" `isSuffixOf` fileName     locationInfo)
    where
        slin = within 100 200
        scol = within   1 100
        elin = within 100 200
        ecol = within   1 100
        within lb ub destructor = let v = destructor locationInfo
                                  in lb <= v && v <= ub

-- (scroll down)
























-- (further down)











































-- XXX: The Line numbers, column numbers and function names matter here!
-- If you move this code, make the appropriate changes in isPeculiar
peculiarFunctionName :: Int -> IO Int
peculiarFunctionName 0 = do
        -- Ok, by now, since we've reached the basecase, we should actually
        -- have some interesting stuff on the executionStack
        stack <- currentExecutionStack
        stackFrames <- getStackFrames stack
        putStrLn $ checkSensibility stackFrames
        return 0
peculiarFunctionName x = IO $ \s ->
    let (IO f) = peculiarFunctionName (x-1)
        (# s', res #) = f s
    in  (# s', res + x #)
      -- By doing like this, we force haskell to put there the update frames
