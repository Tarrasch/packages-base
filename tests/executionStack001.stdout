Found the peculiar function.

15
